backup: switch compactions to distributed SQL processors

kev-cao opened the pull request:
Currently, compactions run locally on the coordinator node. The compaction work should be distributed across the cluster instead. This PR adds distSQL processors and planning logic to compactions. The spans generated by `generateAndSendImportSpans` are divvied up equally across the nodes in the cluster so that each node receives roughly the same number of spans. Progress updates are then sent back to the coordinator node and handled in the same way as the existing backup logic.

This does not include logic for tracing aggregation or checkpointing, which will be done in follow up PRs.

Epic: None

Release note: None

kev-cao committed:
backup: add planning logic for distsql compactions

Currently, compactions run locally on the coordinator node. The
compaction work should be distributed across the cluster instead. This
commit adds the planning logic for how the work should be distributed
across clusters. Note that compaction will still run locally - the
planning logic will be used in a follow up commit that creates the
compaction core processors.

Epic: None

Release note: None

backup: update processor tests

lint

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143769)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice! I only skimmed the PR to look at the distsql bits and left some questions / comments along the way. It doesn't look like a more involved review from Queries is needed, so I'll defer to DR folks.

> Reviewed 6 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143769#-:-OMii75oDatk6lQzxTZr:b-s15lxp)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson and @msbutler)

___
*[`pkg/sql/execinfrapb/processors_bulk_io.proto` line 581 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143769#-OMihvtf4lmL2u0pt302:-OMihvtgAA3gACxOkqXv:b-9u9004) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb904397acfd949cb82ead95db28d49f42fbf879/pkg/sql/execinfrapb/processors_bulk_io.proto#L581)):*
> ```proto
>   optional util.hlc.Timestamp endTime = 6 [(gogoproto.nullable) = false];
>   optional ElidePrefix elide_mode = 7 [(gogoproto.nullable) = false];
>   // spans represents the entire set of spans to be compacted.
> ```

nit: consider expanding the comment to indicate why a particular processor would want to have access to the entire set of spans (i.e. why we're including this information into the spec at all).
___
*[`pkg/backup/compaction_processor_planning.go` line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143769#-OMigiRr2padav6CQ7bx:-OMigiRr2padav6CQ7by:b6p54gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb904397acfd949cb82ead95db28d49f42fbf879/pkg/backup/compaction_processor_planning.go#L48)):*
> ```Go
> 	progCh chan *execinfrapb.RemoteProducerMetadata_BulkProcessorProgress,
> ) error {
> 	defer close(progCh)
> ```

nit: the comment says it's the caller's responsibility to close this channel. Is the comment stale?
___
*[`pkg/backup/compaction_processor_planning.go` line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143769#-OMihF_58R9v7VjSRSTD:-OMihF_58R9v7VjSRSTE:b-f4zej7) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb904397acfd949cb82ead95db28d49f42fbf879/pkg/backup/compaction_processor_planning.go#L158)):*
> ```Go
> 	planCtx, sqlInstanceIDs, err := dsp.SetupAllNodesPlanningWithOracle(
> 		ctx, execCtx.ExtendedEvalContext(), execCtx.ExecCfg(),
> 		physicalplan.DefaultReplicaChooser, roachpb.Locality{},
> ```

nit: do we not want to provide the execution locality?
___
*[`pkg/backup/compaction_processor_planning.go` line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143769#-OMihX8I5jw8MpGLT7xc:-OMihX8J7hlJTbwIGj4l:buoaf6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb904397acfd949cb82ead95db28d49f42fbf879/pkg/backup/compaction_processor_planning.go#L243)):*
> ```Go
> 	}
> 
> 	spanEntryCh := make(chan execinfrapb.RestoreSpanEntry, 1000)
> ```

nit: interesting choice of the size of the channel buffer? What's the rationale for this number? Consider leaving a comment to explain it.


<!-- Sent from Reviewable.io -->

kev-cao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143769)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @jeffswenson, @msbutler, and @yuzefovich)

___
*[`pkg/backup/compaction_processor_planning.go` line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143769#-OMigiRr2padav6CQ7bx:-OMlisSX-aGNZ15xfXDh:b3fyjat) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb904397acfd949cb82ead95db28d49f42fbf879/pkg/backup/compaction_processor_planning.go#L48)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: the comment says it's the caller's responsibility to close this channel. Is the comment stale?
</blockquote></details>

Oh actually it's that line of code that's stale haha, thanks for the catch
___
*[`pkg/backup/compaction_processor_planning.go` line 158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143769#-OMihF_58R9v7VjSRSTD:-OMliySkBUoWdwItsCre:bc828fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb904397acfd949cb82ead95db28d49f42fbf879/pkg/backup/compaction_processor_planning.go#L158)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: do we not want to provide the execution locality?
</blockquote></details>

That's something we will add in a separate PR — there's still work to be done to have compaction support all of the options of `BACKUP`.
___
*[`pkg/backup/compaction_processor_planning.go` line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143769#-OMihX8I5jw8MpGLT7xc:-OMljBGYEdLJjAqAuuhw:b-8hvi69) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb904397acfd949cb82ead95db28d49f42fbf879/pkg/backup/compaction_processor_planning.go#L243)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: interesting choice of the size of the channel buffer? What's the rationale for this number? Consider leaving a comment to explain it.
</blockquote></details>

It's a size that we use in restore, although I don't know if there's a specific rationale for this number. @msbutler @dt do you guys know if there's any empirical reason for this number?
___
*[`pkg/sql/execinfrapb/processors_bulk_io.proto` line 581 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143769#-OMihvtf4lmL2u0pt302:-OMljST62sdEt6LLpkAp:b-d0cxol) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb904397acfd949cb82ead95db28d49f42fbf879/pkg/sql/execinfrapb/processors_bulk_io.proto#L581)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: consider expanding the comment to indicate why a particular processor would want to have access to the entire set of spans (i.e. why we're including this information into the spec at all).
</blockquote></details>

Yeah that sounds good, will update


<!-- Sent from Reviewable.io -->

kev-cao committed:
backup: add distSQL processor for backup compactions

This commit creates a distSQL processor for backup compactions, and
hooks up the existing compaction job logic to use the planning logic
instead of running the compaction locally.

Epic: None

Release note: None

kev-cao commented on pkg/backup/compaction_processor.go:
```diff
@@ -0,0 +1,465 @@
+}
+
+func (p *compactBackupsProcessor) constructProgressProducerMeta(
+	prog execinfrapb.RemoteProducerMetadata_BulkProcessorProgress,
+) *execinfrapb.ProducerMetadata {
+	// Take a copy so that we can send the progress address to the output
+	// processor.
+	progCopy := prog
+	progCopy.NodeID = p.FlowCtx.NodeID.SQLInstanceID()
+	progCopy.FlowID = p.FlowCtx.ID
```
I noticed we did this in `backup_processor.go`, but it seems unnecessary? The value passed to `constructProgressProducerMeta` isn't a pointer, so we could just used the parameter itself.

msbutler replied:
yeah, i actually agree with you. @dt is this a historical artifact?

dt replied:
historical artifact indeed: this code was originally more or less:
```
for prog := range progressCh {
  p := prog
  emit(&p)
}
```

In that construction, we were emitting the _address of_ `p` , rather than its value, and until go1.22, there is only one allocated variable `prog` that has a new value stored in it for each iteration of the loop, so if we'd emitted the address of `prog`, we'd have been emitting the same address over and over, and the receiver of two emitted messages might then read the latest value for both of them.

The copy allocated a new variable in each iteration, the address of which would be unique and safe to emit. In go 1.22, they changes `for` itself to make a new variable for every iteration making the copy pointless, and in some refactors we started passing `prog` by value to a function to do some wrap/unwrap which was making a new copy anyway, so the explicit copy is doubly pointless now.

kev-cao commented on pkg/backup/compaction_processor.go:
```diff
@@ -0,0 +1,465 @@
+			if progCopy.CompletedFraction == nil {
+				progCopy.CompletedFraction = make(map[int32]float32)
+			}
+			progCopy.CompletedFraction[p.ProcessorID] = float32(p.completedSpans) / float32(totalSpans)
```
Curious why we use a map here — can you have more than one processor running per node?

msbutler replied:
heh, i think you can get rid of this. no need to add a `storePerNodeProgressLoop` to compaction for now.

msbutler replied:
i think this is part of some o11y that we never deployed.

kev-cao commented on pkg/backup/compaction_processor.go:
A slight difference here is that in planning, the span covering filter is generated from `manifest.Spans`, whereas `spec.Spans` has had the completed spans removed. The tests didn't seem to have an issue with this, although not sure if that is a problem?

msbutler replied:
we should use the same spans for planning and here. spec.Spans only contains the spans assigned to the proc.

kev-cao replied:
Oh actually `spec.Spans` is the set of all spans that need to be compacted, excluding any that have already been completed (it's what is passed to `generateAndSendImportSpans` in both planning and the processor). `AssignedSpans` is what is assigned to the proc.

msbutler commented on pkg/backup/compaction_processor_planning.go:
could you make this buffer a span group, which will automatically merge adjacent spans? when you bind the spans to `AssignedSpans`, you'll have to convert it back to a good old list of spans. 

(imagine we're sharding up a 100 tb table, we don't want to send 100s of adjacent spans in the spec-- we'd rather just send 1).

kev-cao replied:
That makes a lot of sense, updated!

msbutler commented on pkg/backup/compaction_processor.go:
i think we should use assignSpan.Contains() instead. 

Perhaps as a runtime assertion, we should also error if an entry overlaps with an assigned span but is _not_ contained by it. That would imply that 2 procs are going to work on the same restore span entry, which would lead to sadness.

kev-cao replied:
👍 I do like avoiding sadness

msbutler reviewed (commented):
Really really awesome work here! there's a steep learning curve with this distsql stuff. I have yet to answer your questions or look at your tests.

kev-cao commented on pkg/backup/compaction_job.go:
```diff
@@ -938,18 +679,28 @@ func getBackupChain(
-	)
-	if err != nil {
-		return nil, nil, nil, nil, err
+	// If encryption keys have not already been computed, then we will compute
```
Yeah things break since the processors won't be able to decrypt the files.

kev-cao commented on pkg/sql/execinfrapb/flow_diagram.go:
```diff
@@ -771,6 +771,29 @@ func (i *IngestStoppedSpec) summary() (string, []string) {
 	return "IngestStoppedSpec", []string{detail}
 }
 
+// summary implements the diagramCellType interface.
```
If I ever need a boost of serotonin, I just go to the distSQL view and move those boxes around

kev-cao commented on pkg/backup/compaction_processor.go:
Oh yeah that would be nicer, will do!

kev-cao commented on pkg/backup/compaction_dist_test.go:
```diff
@@ -0,0 +1,123 @@
+		{
+			name: "single node, multiple span entries",
+			entries: []execinfrapb.RestoreSpanEntry{
+				{Span: roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("b")}},
+				{Span: roachpb.Span{Key: roachpb.Key("c"), EndKey: roachpb.Key("d")}},
+				{Span: roachpb.Span{Key: roachpb.Key("e"), EndKey: roachpb.Key("f")}},
+			},
```
TODO: Need to update these tests now that we are using span groups.

kev-cao replied:
Done!

msbutler commented on pkg/backup/compaction_dist.go:
the spans outputted by genSpans are split on size as well?

kev-cao replied:
> // Additionally, if targetSize > 0, we can
// merge the current cover span with the previous cover span if the merged set
// of files have a total data size below the target size.

Oh you're right, well that's nice of it

msbutler commented on pkg/backup/compaction_dist.go:
just noting for the checkpoint PR: i hope we're checkpointing introduced spans correctly.

msbutler commented on pkg/backup/compaction_dist_test.go:
```diff
@@ -0,0 +1,209 @@
+			},
+		},
+		{
+			name: "multiple nodes, single span entry",
```
in the future (no need to change now): for these multi node tests, i think it is sufficient to use 2 nodes, just to reduce the lines of test code, while maintaining the same test coverage.

msbutler commented on pkg/backup/compaction_dist_test.go:
```diff
@@ -0,0 +1,209 @@
+		{
+			name: "single node, multiple span entries",
+			entries: []execinfrapb.RestoreSpanEntry{
+				{Span: roachpb.Span{Key: Key("a"), EndKey: Key("b")}},
```
uber nit: to reduce the chars of code you could use some closures similar to this [test](https://github.com/kev-cao/cockroach/blob/backup/compaction-planning/pkg/crosscluster/physical/stream_ingestion_dist_test.go#L240): so the test case could like: `entries: entries(sp(a,b),sp(c,d)...`

kev-cao replied:
Updated! Looks a lot more readable now for sure

msbutler commented on pkg/backup/compaction_dist_test.go:
```diff
@@ -0,0 +1,209 @@
+		},
+		{
+			name: "multiple nodes, not exact multiple of span entries",
+			entries: []execinfrapb.RestoreSpanEntry{
```
could you use 2 2 1 instead of 3 3 2? just trying to reduce the lines of code a reader needs to read.

msbutler commented on pkg/backup/compaction_dist_test.go:
```diff
@@ -0,0 +1,209 @@
+			},
+		},
+		{
+			name: "multiple nodes, not exact multiple of overlapping span entries",
```
is this combination really adding more coverage?

kev-cao replied:
I suppose not — it did have one nice test cover where some of the spans assigned to a node are overlapped while some are not, but I'll just move that into the other overlap test.

msbutler commented on pkg/backup/compaction_job.go:
```diff
@@ -1016,6 +767,7 @@ func processProgress(
 			manifest.Files = append(manifest.Files, file)
 			manifest.EntryCounts.Add(file.EntryCounts)
 		}
+		// TODO (kev-cao): Add per node progress updates.
```
what is this referring to? i think we just need to persist the intermediate manifests, right?

kev-cao replied:
I noticed that `backup_job` has [progress updates for each node](https://github.com/kev-cao/cockroach/blob/673c8c1796855bf2f729a1f9106a24217713d3d2/pkg/backup/backup_job.go#L231-L244), that seemed like a nice to have for compactions too.

msbutler commented on pkg/backup/compaction_processor.go:
mark this as an assertion failure?

msbutler reviewed (commented):
Nearly there!

msbutler reviewed (approved):
er, my nits are non blocking.

kev-cao commented:
TFTR!

bors r=msbutler
